sql/semisync_master.cc

```diff
@@ -705,16 +673,19 @@ int Repl_semi_sync_master::report_reply_binlog(uint32 server_id,
 
   if (need_copy_send_pos)
   {
+    Tranx_node *entry;
     strmake_buf(m_reply_file_name, log_file_name);
     m_reply_file_pos = log_file_pos;
     m_reply_file_name_inited = true;
 
-    /* Remove all active transaction nodes before this point. */
-    DBUG_ASSERT(m_active_tranxs != NULL);
-    m_active_tranxs->clear_active_tranx_nodes(log_file_name, log_file_pos,
-                                              signal_waiting_transaction);
-    if (m_active_tranxs->is_empty())
-      m_wait_file_name_inited= false;
+    entry= m_active_tranxs->get_tranx_node(log_file_name, log_file_pos);
+    if (entry && ++(entry->acks) >= rpl_semi_sync_master_wait_for_slave_count)
+    {
+      /* Remove all active transaction nodes before this point. */
+      m_active_tranxs->clear_active_tranx_nodes(entry->next);
+      if (m_active_tranxs->is_empty())
+        m_wait_file_name_inited= false;
+    }
```

Comment:
I think now that we could need multiple ACKs, this logic here which does the
actual clearing and signalling probably shouldn't be grouped in the
need_copy_send_pos conditional. I imagine that if there are a couple
transactions actively awaiting acks, and one slave is fast, and another is
slow, that when handling ACKs from the slow slave, need_copy_send_pos would
be false because m_reply_file_{pos,name} would already be updated from the
fast slave. It'd be good to have a test case for this scenario too.

Reply:
To fix the “slow replica” bug, I moved the `m_reply_file_{pos,name}` updates to (a reüsable method that’s called) under the `if (entry …)` condition.
Now, there is practically nothing between `if (need_copy_send_pos)` and this `if`.

Although I think, when `entry= …->get_tranx_node(…)` exists, `need_copy_send_pos` would always be true.
I think the converse is only false when this code is called from `Repl_semi_sync_master::dump_start()`.
Overall, I think the `m_reply_file_{name,pos}` variables are now only significant for the states of `is_no_slave()` that flush transactions without updating this position.

`Repl_semi_sync_master::report_reply_binlog()` might have an identity crisis if we have more to refactor away.